kubeadm panic on CoreDNS upgrade

Latest vanilla master crashes upon upgrade

```
root@kind-control-plane:/# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.0-alpha.1.15+894916b66559e5", GitCommit:"894916b66559e588b32c1ac803232afaa850ed68", GitTreeState:"clean", BuildDate:"2020-03-26T17:05:27Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
root@kind-control-plane:/# kubeadm upgrade apply --allow-experimental-upgrades v1.19.0-alpha.1
...
panic: runtime error: index out of range [2] with length 2

goroutine 1 [running]:
k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns.isCoreDNSVersionSupported(0x1c9b560, 0xc0000c94a0, 0xc000386780, 0xc0008d4280, 0x130)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns/dns.go:414 +0x2ad
k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns.createCoreDNSAddon(0xc0003a5800, 0x7fb, 0xb5f, 0xc000238000, 0x309, 0x418, 0xc000507800, 0x1df, 0x3b1, 0x1c9b560, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns/dns.go:276 +0x17d
k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns.coreDNSAddon(0xc0005b9d60, 0x1c9b560, 0xc0000c94a0, 0xc00038f58c, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns/dns.go:257 +0x5ad
k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns.EnsureDNSAddon(0xc0005b9d60, 0x1c9b560, 0xc0000c94a0, 0x0, 0xc000337700)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns/dns.go:109 +0x139
k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade.PerformPostUpgradeTasks(0x1c9b560, 0xc0000c94a0, 0xc0005b9d40, 0xc000810e40, 0xc00066a000, 0xc0005b9d40, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade/postupgrade.go:99 +0x46f
k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade.runApply(0xc0000eeb70, 0x7ffca2d1a94f, 0xf, 0x2, 0x1)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade/apply.go:170 +0x556
k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade.NewCmdApply.func1(0xc000406280, 0xc0000bc660, 0x1, 0x2, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade/apply.go:79 +0xa4
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc000406280, 0xc0000bc5c0, 0x2, 0x2, 0xc000406280, 0xc0000bc5c0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:826 +0x460
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc0000a8a00, 0xc00000e010, 0x1c0f680, 0xc00000e018)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:914 +0x2fb
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:864
k8s.io/kubernetes/cmd/kubeadm/app.Run(0x0, 0x16ed780)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50 +0x21c
main.main()
        _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25 +0x22
```

```
   containerStatuses:
    - containerID: containerd://3ef85d22e9be0090b02545310a3c25a71f805774aca0feaa1450aeb911f99cdd
      image: k8s.gcr.io/coredns:1.6.7
      imageID: sha256:67da37a9a360e600e74464da48437257b00a754c77c40f60c65e4cb327c34bd5
```

/kind bug
/priority critical-urgent
/assign

--- a/cmd/kubeadm/app/phases/addons/dns/dns.go
+++ b/cmd/kubeadm/app/phases/addons/dns/dns.go
@@ -21,6 +21,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"net"
+	"regexp"
 	"strings"
 
 	"github.com/caddyserver/caddy/caddyfile"
@@ -397,6 +398,11 @@ func isCoreDNSConfigMapMigrationRequired(corefile, currentInstalledCoreDNSVersio
 	return isMigrationRequired, nil
 }
 
+var (
+	// imageDigestMatcher is used to match the SHA256 digest from the ImageID of the CoreDNS pods
+	imageDigestMatcher = regexp.MustCompile(`^.*(?i:sha256:([[:alnum:]]{64}))$`)
+)
+
 func isCoreDNSVersionSupported(client clientset.Interface) (bool, error) {
 	isValidVersion := true
 	coreDNSPodList, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(
@@ -410,8 +416,12 @@ func isCoreDNSVersionSupported(client clientset.Interface) (bool, error) {
 	}
 
 	for _, pod := range coreDNSPodList.Items {
-		imageID := strings.Split(pod.Status.ContainerStatuses[0].ImageID, ":")
-		if !migration.Released(imageID[2]) {
+		imageID := imageDigestMatcher.FindStringSubmatch(pod.Status.ContainerStatuses[0].ImageID)
+		if len(imageID) != 2 {
+			return false, errors.Errorf("unable to match SHA256 digest ID in %q", pod.Status.ContainerStatuses[0].ImageID)
+		}
+		// The actual digest should be at imageID[1]
+		if !migration.Released(imageID[1]) {
 			isValidVersion = false
 		}
 	}

